gh-82012: Deprecate bitwise inversion (~) of bool (#103487)
python/cpython: Doc/library/stdtypes.rst

           hash_value = -2
       return hash_value

.. _typebool:

Boolean Type - :class:`bool`
============================

Booleans represent truth values. The :class:`bool` type has exactly two

> constant instances ``True`` and ``False``.

Reviewer: ```suggestion
constant instances: ``True`` and ``False``.
```

Author: done.

> Other types can be cast to bool explicitly using the function :func:`bool`.

Reviewer: I think I preferred the old phrasing:

> The built-in function [bool()](https://docs.python.org/3/library/functions.html#bool) can be used to convert any value to a Boolean, if the value can be interpreted as a truth value (see section [Truth Value Testing](https://docs.python.org/3/library/stdtypes.html#truth) above)

It draws a clearer distinction between types and values and sidesteps the issue that `and` and `or` don't actually convert their operands to bool.

Author: don.

Additionally, they are implicitly interpreted as bool in some contexts - see
:ref:`truth`.

For logical operations, use the :ref:`boolean operators <boolean>` ``and``,
``or`` and ``not``.
When applying the bitwise operators ``&``, ``|``, ``^`` to two booleans, they
return a bool equivalent to the logical operations "and", "or", "xor". Still,
for ``&`` and ``|``, the logical operators ``and`` and ``or`` are preferred.

> The use of the bitwise inversion operator ``~`` is deprecated and will raise

Reviewer: Maybe use the `.. deprecated::` role to mention the deprecation

Author: done.

           hash_value = -2
       return hash_value

.. _typebool:

Boolean Type - :class:`bool`
============================

Booleans represent truth values. The :class:`bool` type has exactly two
constant instances: ``True`` and ``False``.

.. index::
   single: False
   single: True
   pair: Boolean; values

The built-in function :func:`bool`  converts any value to a boolean, if the
value can be interpreted as a truth value (see section :ref:`truth` above).

For logical operations, use the :ref:`boolean operators <boolean>` ``and``,
``or`` and ``not``.
When applying the bitwise operators ``&``, ``|``, ``^`` to two booleans, they

> return a bool equivalent to the logical operations "and", "or", "xor". Still,
> for ``&`` and ``|``, the logical operators ``and`` and ``or`` are preferred.

Reviewer: (ignorable) How do you feel about suggesting an alternative for xor as well?
```suggestion
return a bool equivalent to the logical operations "and", "or", "xor". However,
the logical operators ``and``, ``or`` and ``!=`` should be preferred
over ``&``, ``|`` and ``^``.
```

.. deprecated:: 3.12

   The use of the bitwise inversion operator ``~`` is deprecated and will
   raise an error in Python 3.14.

:class:`bool` is a subclass of :class:`int` (see :ref:`typesnumeric`). In

> many numeric contexts, bools behave like the integers 0 and 1, respectively.

Reviewer: So that the "respectively" makes sense
```suggestion
many numeric contexts, ``False`` and ``True`` behave like the integers 0 and 1, respectively.
```

It is written as ``NotImplemented``.

> .. _bltin-boolean-values:

Reviewer 2: Hello. The removal of this `ref` tag broke intersphinx downstream. I have not tracked down yet which middleware is calling this tag when API docstring has a "bool" mentioned. I suspect it is numpydoc. Is it not possible to reuse this tag for your new section above?

xref https://github.com/astropy/astropy/issues/15428

Reviewer: If you open a PR, I can backport it to the 3.12 branch

Reviewer 2: Sure. Please see https://github.com/python/cpython/pull/110371 . Thank you for your consideration!
